build/archives/BUILD.gn

```diff
@@ -6,7 +6,7 @@ import("//flutter/build/zip_bundle.gni")
 import("//flutter/common/config.gni")
 
 if (build_engine_artifacts) {
-  zip_bundle("artifacts") {
+  zip_bundle_with_entitlements("artifacts") {
```

Comment:
Yeah the same pattern is needed for multiple zip bundles, which live in different BUILD.gn files, such as andriod, mac, tools etc. I put the logic in a template to save some typings.

Yeah I only included artifacts.zip in this commit. This is based on the feedback to start with a single zip, and smaller pr for easy review. (https://github.com/flutter/engine/pull/34112#pullrequestreview-1011235510)

Reply:
I see. I think it could still be a bit cleaner to factor this a little differently instead of using a special purpose `zip_...` template. One idea that I'm not 100% about would be to make a template specifically for the metadata files. That template would accept the same file list-of-scopes that is passed to the `zip_bundle` template. You can set additional values in those scopes that indicate whether the files are "with" or "without" entitlements:
```gn
files += [
  {
    source = "//third_party/icu/flutter/icudtl.dat"
    destination = "icudtl.dat"
    entitlement = "none" # neither "with" nor "without"
  },
  {
    source = "$root_out_dir/flutter_tester$exe"
    destination = "flutter_tester$exe"
    entitlement = "with" # or "without"
  },
  ...
```

With a template like:
```gn
template("write_entitlements") {
  with = []
  without = []
  foreach(file, invoker.files) {
    if (file.entitlement == "with") {
      with += [...]
    } else if (file.entitlement == "without") {
      without += [...]
    }
  }
  write_file(..., with)
  write_file(..., without)
  # Might need a copy() target here to make this a real template from which you can capture the outputs.
}
```

Then in the `zip_bundle` rule, you can depend on the `write_entitlements` target, and extract its outputs with `get_target_output`.

Reply:
umm I might be missing something, but my understanding is that an engine artifact can have many files (binary and non-binary files) inside a zip. With the proposed approach, we might need to add an `entitlement = none` field to 16 non-binary files in Flutter.framework/Headers for example, and to many other non-binary files. This could be a lot of changes. 

I think the pros of using an array of entitlements file paths is that, we only specify a list of the affected binaries. The affected binaries are roughly at the quantity of 1~4 ish per zip file, and change wise it is a few extra lines of code per zip file. I could be wrong, but I think this way it is cleaner.

Reply:
Discussed with zra offline and the current plan is to collect metadata through the `generated_file` api.

I am currently looking at how to make `generated_file` target work with the `zip_bundle` template. Because they can't be nested, and system would complain `You are trying to nest a target inside a template invocation.`